File: tests/Circuit.FSharp.Tests/DocumentationExamples/StructuredOutput.fs
namespace Circuit.FSharp.Tests.DocumentationExamples

open System.ComponentModel.DataAnnotations
open System.Threading
open Circuit.Core
open Circuit.FSharp

[<AllowNullLiteral>]
type ContactRequest() =
    [<property: Required>]
    member val Message = "" with get, set

[<AllowNullLiteral>]
type ContactSummary() =
    [<property: Required>]
    member val Name = "" with get, set

    [<property: Required>]
    member val Status = "" with get, set

module StructuredOutputExample =
    let signature =
        Signature.create<ContactRequest, ContactSummary>
            "contacts.signature"
            "1.0.0"
            "Contact summary"
            "Return name and status."

    let run (runtime: ICircuitRuntime) (agent: AgentDefinition) =
        Agent.run
            runtime
            agent
            signature
            (ContactRequest(Message = "Summarize Ada Lovelace."))
            RunOptions.Default
            CancellationToken.None

File: tests/Circuit.Interop.Tests/DocumentationExamples/StructuredOutput.cs
using System.ComponentModel.DataAnnotations;
using Circuit;
using Microsoft.Extensions.AI;

namespace Circuit.Interop.Tests.DocumentationExamples;

internal static class StructuredOutputExample
{
    public static Task<AgentRunResult<ContactSummary>> RunWithNativeOnlyAsync(
        ICircuitClient client,
        CancellationToken cancellationToken)
    {
        var options = new AgentRunOptions
        {
            StructuredOutputPolicy = StructuredOutputPolicy.NativeOnly,
        };

        return client.RunAsync(
            new AgentDefinition("contacts.agent", "1.0.0", "Contacts", "Return only structured JSON."),
            new AgentSignature<ContactRequest, ContactSummary>("contacts.signature", "1.0.0", "Contact summary", "Return name and status."),
            new ContactRequest { Message = "Summarize Ada Lovelace." },
            options,
            cancellationToken);
    }

    public static ICircuitClient BuildRepairingClient(IChatClient primaryClient, IChatClient repairClient)
        => new CircuitClientBuilder()
            .UseMicrosoftAgentFramework(primaryClient)
            .ConfigureMicrosoftAgentFramework(options => options.SecondaryStructuredOutputClient = repairClient)
            .Build();

    internal sealed class ContactRequest
    {
        [Required]
        public string Message { get; set; } = string.Empty;
    }

    internal sealed class ContactSummary
    {
        [Required]
        public string Name { get; set; } = string.Empty;

        [Required]
        public string Status { get; set; } = string.Empty;
    }
}
